Add Trino 440 release notes (#20797)
trinodb/trino: docs/src/main/sphinx/release/release-440.md

# Release 440 (21 Feb 2024)

## Delta Lake connector

> * Add support for concurrent writes with `INSERT` statements. ({issue}`18506`)

Reviewer: Maybe:

"Add support for concurrent `INSERT` queries"

## Hive connector

> * Add support for using an HTTP client with the Thrift metastore. ({issue}`20371`)

Reviewer 2: do we have docs for that ? Update in https://trino.io/docs/current/admin/properties-http-client.html ? and elsewhere?

Author: It's documented alongside the Thrift metastore properties, yeah.

> * Add support for using an HTTP client with the Thrift metastore. ({issue}`20371`)

Reviewer 2: Maybe

```suggestion
* Add support for bearer token authentication for Thrift metastore connection. ({issue}`20371`)
```

Also https://github.com/trinodb/trino/pull/20799

> * Add support for commenting on partitioned columns in the Thrift metastore. ({issue}`20264`)

Reviewer 2: whoa.. really .. partitioned is significant here?

Author: Yep!

Reviewer 2: asked in PR

> * Add support for `varchar` to `float` coercion. ({issue}`20719`)

Reviewer 2: in general or in schema evolution?

* Add support for changing a column's type from `varchar` to `char`. ({issue}`20723`)
* Add support for changing a column's type from `varchar` to `boolean`. ({issue}`20741`)
* Improve performance of reading from JSON files. ({issue}`19396`)

> * Fix incorrect truncation when decoding `varchar` and `char(n)` in `TEXTFILE`

Reviewer 3: `varchar` -> `varchar(n)`

  and `SEQUENCEFILE` formats. ({issue}`20731`)

## Iceberg connector

* Disallow setting the materialized view owner when using system security with
  the Glue catalog. ({issue}`20647`)

## SPI

* Add reset to position method to `BlockBuilder`. ({issue}`19577`)
* Remove the `getChildren` method from `Block`. ({issue}`19577`)

> * Remove the `get{Type}` methods from `Block`.  Callers should unwrap a `Block`

Reviewer 2: Must vs should ?

> # Release 440 (21 Feb 2024)

Reviewer 2: 23

## Web UI

* Fix error when using authentication tokens larger than 4 kB. ({issue}`20787`)

## Delta Lake connector

* Add support for concurrent `INSERT` queries. ({issue}`18506`)

## Hive connector

* Add support for bearer token authentication for a Thrift metastore connection. ({issue}`20371`)
* Add support for commenting on partitioned columns in the Thrift metastore. ({issue}`20264`)

> * Add support for changing a column's type from `varchar` to `float`. ({issue}`20719`)

Reviewer 2: maybe 

```suggestion
* Add support for changing column type from `varchar` to `float`. ({issue}`20719`)
```

also in others

Author: Not sure I'm a fan of this change, takes it from proper grammar to improper

Reviewer 2: sounds good

* Add support for changing a column's type from `varchar` to `char`. ({issue}`20723`)
* Add support for changing a column's type from `varchar` to `boolean`. ({issue}`20741`)
* Improve performance of reading from JSON files. ({issue}`19396`)
* Fix incorrect truncation when decoding `varchar(n)` and `char(n)` in
  `TEXTFILE` and `SEQUENCEFILE` formats. ({issue}`20731`)

> * Fix query failure when `hive.file-status-cache-tables` is enabled for a table

Reviewer 2: no comma anywhere?

Author: I don't think so...

  and new manifest files have been added but not cached yet. ({issue}`20344`)

## Iceberg connector

> * Improve latency of queries when filesystem caching is enabled. ({issue}`20803`)

Reviewer 2: ```suggestion
* Improve latency of queries when file system caching is enabled. ({issue}`20803`)
```

# Release 440 (7 Mar 2024)

## General

* Add [Snowflake connector](/connector/snowflake). ({issue}`17909`)
* Add support for sub-queries inside `UNNEST` clauses. ({issue}`17953`)
* Improve performance of {func}`arrays_overlap`. ({issue}`20900`)
* Export JMX statistics for resource groups by default. This can be disabled
  with the `jmxExport` resource group property. ({issue}`20810`)

> * Fix query failure when a `CHECK` constraint is null. ({issue}`20906`)

Reviewer 2: not sure if it should use `CHECK` or just check constraint

* Fix query failure for aggregations over `CASE` expressions when the input
  evaluation could throw an error. ({issue}`20652`)

> * Fix incorrect results when running a SQL routine with a single if/else

Reviewer 2: Fix incorrect behavior of the `else` clause in a SQL routines with a single if/end condition.

Or 

Prevent ignoring else clause...

  condition. ({issue}`20926`)

> * Fix `OPTIMIZE` queries failing due to exceeding the open writer limit. ({issue}`20871`)

Reviewer 2: do we need to say ALTER TABLE EXECUTE optimize ? or call it a table procedure instead of a query?

* Fix certain `INSERT` and `CREATE TABLE AS .. SELECT` queries failing due to
  exceeding the of open writer limit on partitioned tables. ({issue}`20871`)
* Fix "multiple entries with same key" query failure for queries with joins on
  partitioned tables. ({issue}`20917`)

## Security

> * Add support for row filtering and column masking in OPA access control. ({issue}`20921`)

Reviewer 2: maybe Open Policy Agent access control ?

Reviewer 2: ping @colebow ?

## Web UI

* Fix error when using authentication tokens larger than 4 kB. ({issue}`20787`)

> * Fix incorrect results when using `GRANT`, `DENY`, and `REVOKE` clauses on

Reviewer 2: Why is that in the Web UI section?

  views and materialized views. ({issue}`20812`)

## Delta Lake connector

* Add support for concurrent `INSERT` queries. ({issue}`18506`)

> * Improve latency for queries with filesystem caching enabled. ({issue}`20851`)

Reviewer 2: file system caching

* Improve latency for queries on tables with checkpoints. ({issue}`20901`)

## Hive connector

* Add support for bearer token authentication for a Thrift metastore connection. ({issue}`20371`)
* Add support for commenting on partitioned columns in the Thrift metastore. ({issue}`20264`)
* Add support for changing a column's type from `varchar` to `float`. ({issue}`20719`)
* Add support for changing a column's type from `varchar` to `char`. ({issue}`20723`)
* Add support for changing a column's type from `varchar` to `boolean`. ({issue}`20741`)
* Add support for configuring a `region` and `endpoint` for S3 security mapping. ({issue}`18838`)
* Improve performance when reading JSON files. ({issue}`19396`)
* Fix incorrect truncation when decoding `varchar(n)` and `char(n)` in
  `TEXTFILE` and `SEQUENCEFILE` formats. ({issue}`20731`)
* Fix query failure when `hive.file-status-cache-tables` is enabled for a table
  and new manifest files have been added but not cached yet. ({issue}`20344`)
* Fix error when trying to `INSERT` into a transactional table that does not
  have partitions. ({issue}`19407`)

## Iceberg connector

> * Improve latency of queries when file system caching is enabled. ({issue}`20803`)

Reviewer 2: hahah .. here its right...

* Disallow setting the materialized view owner when using system security with
  the Glue catalog. ({issue}`20647`)

> * Rename the `orc.bloom.filter.columns` and `orc.bloom.filter.fpp` table

Reviewer 2: Breaking change? Also .. are docs updated?

Author: Old names are deprecated but still functional, so not a breaking change. Properties were never documented in the first place 😶

Reviewer 2: Can you send a doc PR

# Release 440 (7 Mar 2024)

## General

* Add support for sub-queries inside `UNNEST` clauses. ({issue}`17953`)
* Improve performance of {func}`arrays_overlap`. ({issue}`20900`)

> * Change the default value for the `jmxExport` resource group property to

Reviewer: "Export JMX stats for resource groups by default. This can be turned off via the xxx property"

  `true`. ({issue}`20810`)
* Fix query failure when a `CHECK` constraint is null. ({issue}`20906`)

> * Fix query failure for `CASE` aggregation queries when the input evaluation

Reviewer: There's no such a thing as a "CASE aggregation query". Replace it with something like "aggregations over conditional expressions" or "aggregations over CASE expressions".

  partitioned tables. ({issue}`20917`)

## Security

* Add support for row filtering and column masking in OPA access control. ({issue}`20921`)

## Web UI

* Fix error when using authentication tokens larger than 4 kB. ({issue}`20787`)
* Fix incorrect results when using `GRANT`, `DENY`, and `REVOKE` clauses on
  views and materialized views. ({issue}`20812`)

## Delta Lake connector

* Add support for concurrent `INSERT` queries. ({issue}`18506`)
* Improve latency for queries with filesystem caching enabled. ({issue}`20851`)
* Improve latency for queries on tables with checkpoints. ({issue}`20901`)

## Hive connector

* Add support for bearer token authentication for a Thrift metastore connection. ({issue}`20371`)
* Add support for commenting on partitioned columns in the Thrift metastore. ({issue}`20264`)
* Add support for changing a column's type from `varchar` to `float`. ({issue}`20719`)
* Add support for changing a column's type from `varchar` to `char`. ({issue}`20723`)
* Add support for changing a column's type from `varchar` to `boolean`. ({issue}`20741`)
* Add support for configuring a `region` and `endpoint` for S3 security mapping. ({issue}`18838`)

> * Improve performance of reading from JSON files. ({issue}`19396`)

Reviewer: I think "when reading JSON files" would sound better.

* Fix incorrect truncation when decoding `varchar(n)` and `char(n)` in
  `TEXTFILE` and `SEQUENCEFILE` formats. ({issue}`20731`)
* Fix query failure when `hive.file-status-cache-tables` is enabled for a table
  and new manifest files have been added but not cached yet. ({issue}`20344`)

> * Fix error when trying to `INSERT` into an ACID table that does not have

Reviewer: "into a *transactional* table"

  partitions. ({issue}`19407`)

## Iceberg connector

* Improve latency of queries when file system caching is enabled. ({issue}`20803`)
* Disallow setting the materialized view owner when using system security with
  the Glue catalog. ({issue}`20647`)
* Rename the `orc.bloom.filter.columns` and `orc.bloom.filter.fpp` table
  properties to `write.orc.bloom.filter.columns` and
  `write.orc.bloom.filter.fpp`, respectively. ({issue}`20432`)

## Snowflake connector

> * Add [Snowflake connector](/connector/snowflake). ({issue}`17909`)

Reviewer: This should go in the general section. Future release note entries that affect the connector will go in this section.
